### ISSUE TITLE
Rename Linear to Affine.

### ISSUE THREAD

**Issue by dmitriy-serdyuk** [mila-udem/blocks#734]
For details see discussion in #716.

**Comment by dwf** [comment]
This is one place where machine learning/neural networks jargon is squarely
on the side of the "incorrect" name Linear, I'm afraid. Thus this rename is
probably not such a great idea.

**Comment by rizar** [comment]
Isn't it an exact copy of the situation with `Sigmoid` vs `Logistic` ? ;)

**Comment by dwf** [comment]
Nah, the logistic function is well known by that name (even has a Wikipedia
page).

Also, if we are to introduce a Linear layer equivalent to the current
Linear(use_bias=False) this will silently alter the operation of a lot of
user code.

**Comment by rizar** [comment]
Affine function also have a wikipedia web-page. But your second argument is quite a strong one... 

Let's see what @bartvm thinks when he is back.

**Comment by dwf** [comment]
To clarify my first argument, the logistic vs sigmoid thing is a matter of
precision: sigmoid simply means S-shaped, and the logistic is one of
several S-shaped functions in use. The inheritance goes in the other
direction for linear and affine: all linear transformations are a special
case of affine transformations but not vice versa. But the main thrust is
that people have been calling affine layers "linear layers" in the neural
networks world for at least a decade, and so the nomenclature, while
incorrect, is probably here to stay.

That said, I believe the "silent removal of biases" argument is a much
bigger practical concern.

**Comment by dmitriy-serdyuk** [comment]
We can just get rid of `Linear` and rename it to `Affine`. This is going to be better from mathematical point of view and users can still use `Affine(use_bias=False)`.

**Comment by dwf** [comment]
Mmm, I still think we will have a lot of people asking how to create a
"linear" layer on the mailing list/bug tracker/etc., as that's the more
common name...
